src/main/java/io/supertokens/webserver/api/bulkimport/AddBulkImportUsers.java

```diff
@@ -0,0 +1,170 @@
+import io.supertokens.pluginInterface.multitenancy.AppIdentifier;
+import io.supertokens.pluginInterface.multitenancy.AppIdentifierWithStorage;
+import io.supertokens.pluginInterface.multitenancy.TenantConfig;
+import io.supertokens.pluginInterface.multitenancy.TenantIdentifier;
+import io.supertokens.pluginInterface.multitenancy.exceptions.TenantOrAppNotFoundException;
+import io.supertokens.utils.Utils;
+import io.supertokens.webserver.InputParser;
+import io.supertokens.webserver.WebserverAPI;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+
+public class AddBulkImportUsers extends WebserverAPI {
+    private static final int MAX_USERS_TO_ADD = 10000;
+
+    public AddBulkImportUsers(Main main) {
+        super(main, "");
+    }
+
+    @Override
+    public String getPath() {
+        return "/bulk-import/add-users";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        JsonArray users = InputParser.parseArrayOrThrowError(input, "users", false);
+
+        if (users.size() <= 0 || users.size() > MAX_USERS_TO_ADD) {
+            JsonObject errorResponseJson = new JsonObject();
+            String errorMsg = users.size() <= 0 ? "You need to add at least one user."
+                    : "You can only add 10000 users at a time.";
+            errorResponseJson.addProperty("error", errorMsg);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
+        }
+
+        AppIdentifierWithStorage appIdentifierWithStorage;
+        try {
+            appIdentifierWithStorage = getAppIdentifierWithStorageFromRequestAndEnforcePublicTenant(req);
+        } catch (TenantOrAppNotFoundException | BadPermissionException e) {
+            throw new ServletException(e);
+        }
+
+        JsonArray errorsJson = new JsonArray();
+        ArrayList<BulkImportUser> usersToAdd = new ArrayList<>();
+
+        for (int i = 0; i < users.size(); i++) {
+            try {
+                BulkImportUser user = BulkImportUserUtils.createBulkImportUserFromJSON(users.get(i).getAsJsonObject(), Utils.getUUID());
+                
+                for (BulkImportUser.LoginMethod loginMethod : user.loginMethods) {
+                    validateTenantId(appIdentifierWithStorage, loginMethod.tenantId, loginMethod.recipeId);
+
+                    if (loginMethod.emailPasswordLoginMethod != null) {
+                        validatePasswordHashingAlgorithm(appIdentifierWithStorage, loginMethod.emailPasswordLoginMethod);
+                    }
```

Comment:
both of these should be in the `createBulkImportUserFromJSON` function?

Reply:
done